src/main/java/me/ryanhamshire/GriefPrevention/Claim.java

```diff
@@ -443,214 +395,184 @@ public boolean hasExplicitPermission(Player player, ClaimPermission level)
-    public String allowAccess(Player player)
-    {
-        //following a siege where the defender lost, the claim will allow everyone access for a time
-        if (this.doorsOpen) return null;
+    private String getPermissionMessage(Player player, ClaimPermission permission, Event event) {
+        // If we don't know who's asking, always say no.
+        if (player == null) return "";
 
-        //admin claims need adminclaims permission only.
+        // Admin claims need adminclaims permission only.
         if (this.isAdminClaim())
         {
             if (player.hasPermission("griefprevention.adminclaims")) return null;
         }
 
-        //claim owner and admins in ignoreclaims mode have access
+        // Anyone with deleteclaims permission can edit non-admin claims at any time.
+        else if (permission == ClaimPermission.Edit && player.hasPermission("griefprevention.deleteclaims"))
+            return null;
+
+        // Claim owner and admins in ignoreclaims mode have access.
         if (player.getUniqueId().equals(this.ownerID) || GriefPrevention.instance.dataStore.getPlayerData(player.getUniqueId()).ignoreClaims)
             return null;
 
-        //look for explicit individual access, inventory, or build permission
+        // Look for explicit individual permission.
         if (this.hasExplicitPermission(player, ClaimPermission.Access)) return null;
 
-        //also check for public permission
-        if (ClaimPermission.Access.isGrantedBy(this.playerIDToClaimPermissionMap.get("public"))) return null;
-
-        //permission inheritance for subdivisions
-        if (this.parent != null)
-        {
-            if (player.getUniqueId().equals(this.parent.ownerID))
-                return null;
-            if (!inheritNothing)
-                return this.parent.allowAccess(player);
-        }
-
-        //catch-all error message for all other cases
-        String reason = GriefPrevention.instance.dataStore.getMessage(Messages.NoAccessPermission, this.getOwnerName());
-        if (player.hasPermission("griefprevention.ignoreclaims"))
-            reason += "  " + GriefPrevention.instance.dataStore.getMessage(Messages.IgnoreClaimsAdvertisement);
-        return reason;
-    }
-
-    //inventory permission check
-    public String allowContainers(Player player)
-    {
-        //if we don't know who's asking, always say no (i've been told some mods can make this happen somehow)
-        if (player == null) return "";
+        // Check for public permission.
+        if (permission.isGrantedBy(this.playerIDToClaimPermissionMap.get("public"))) return null;
 
-        //trying to access inventory in a claim may extend an existing siege to include this claim
-        GriefPrevention.instance.dataStore.tryExtendSiege(player, this);
-
-        //if under siege, nobody accesses containers
-        if (this.siegeData != null)
+        // Special building-only rules.
+        if (permission == ClaimPermission.Build)
```

Comment:
Not a huge fan of special casing a ClaimPermission, but I thought that this was a better choice than another listener specifically to modify behavior in this case. Farming and PVP could be candidates for addons like Siege, but they're a lot lower priority so I kept the logic inside the claim.